### ISSUE TITLE
Compiling with 3.14 Vanilla Kernel fails

### ISSUE THREAD

**Issue by thorstenkramm** [schoebel/mars#9]
Hello,

I tried to install and test mars. But with no luck so far.
My system is a recent virgin Debian Jessie 64bit.
I tried to install mars into a vanilla kernel 3.14.58. 
Applying the patch `vanilla-3.14/0001-mars-SPECIAL-for-in-tree-build.patch` from the master branch worked flawless.
But compiling the kernel modules failed with the following error:
```
MARS: Generating block/mars/kernel/buildtag.h
cat block/mars/kernel/buildtag.h;
/* Automatically generated -- DO NOT EDIT! */
#define BUILDTAG  "light0.1stable23"
#define BUILDHOST "root@debian"
#define BUILDDATE "2016-01-22 21:43:36"
  CC [M]  block/mars/kernel/lamport.o
  CC [M]  block/mars/kernel/brick_say.o
  CC [M]  block/mars/kernel/brick_mem.o
  CC [M]  block/mars/kernel/brick.o
  CC [M]  block/mars/kernel/mars_generic.o
  CC [M]  block/mars/kernel/lib_log.o
  CC [M]  block/mars/kernel/lib_rank.o
  CC [M]  block/mars/kernel/lib_limiter.o
  CC [M]  block/mars/kernel/lib_timing.o
  CC [M]  block/mars/kernel/lib_mapfree.o
  CC [M]  block/mars/kernel/mars_net.o
  CC [M]  block/mars/kernel/mars_server.o
  CC [M]  block/mars/kernel/mars_client.o
  CC [M]  block/mars/kernel/mars_aio.o
  CC [M]  block/mars/kernel/mars_sio.o
  CC [M]  block/mars/kernel/mars_bio.o
  CC [M]  block/mars/kernel/mars_if.o
block/mars/kernel/mars_if.c: In function ‘if_switch’:
block/mars/kernel/mars_if.c:881:17: error: ‘MARS_MAJOR’ undeclared (first use in this function)
   disk->major = MARS_MAJOR; //TODO: make this dynamic for >256 devices
                 ^
block/mars/kernel/mars_if.c:881:17: note: each undeclared identifier is reported only once for each function it appears in
block/mars/kernel/mars_if.c: In function ‘exit_mars_if’:
block/mars/kernel/mars_if.c:1301:20: error: ‘MARS_MAJOR’ undeclared (first use in this function)
  unregister_blkdev(MARS_MAJOR, "mars");
                    ^
block/mars/kernel/mars_if.c: In function ‘init_mars_if’:
block/mars/kernel/mars_if.c:1311:27: error: ‘MARS_MAJOR’ undeclared (first use in this function)
  status = register_blkdev(MARS_MAJOR, "mars");
                           ^
scripts/Makefile.build:308: recipe for target 'block/mars/kernel/mars_if.o' failed
make[4]: *** [block/mars/kernel/mars_if.o] Error 1
scripts/Makefile.build:455: recipe for target 'block/mars/kernel' failed
make[3]: *** [block/mars/kernel] Error 2
Makefile:845: recipe for target 'block' failed
make[2]: *** [block] Error 2
/root/kernel/linux-3.14.58/scripts/package/Makefile:90: recipe for target 'deb-pkg' failed
make[1]: *** [deb-pkg] Error 2
Makefile:1131: recipe for target 'deb-pkg' failed
make: *** [deb-pkg] Error 2
```
Am I doing something wrong? Or can somebody give me hint, how to install mars?

Bellow are the exact steps I've done:
```bash
cd /usr/src
wget https://cdn.kernel.org/pub/linux/kernel/v3.x/linux-3.14.58.tar.xz
mkdir ~/kernel
cd ~/kernel
tar -xaf /usr/src/linux-3.14.58.tar.xz
cd ~/kernel/linux-3.14.58/block
git clone git://github.com/schoebel/mars
cd ~/kernel/linux-3.14.58
patch -p1 < block/mars/pre-patches/vanilla-3.14/0001-mars-SPECIAL-for-in-tree-build.patch
make olddefconfig
make menuconfig
```
This gave no errors.
I enabled Mars and started to build the kernel with `make -j6 deb-pkg`.

Thanks a lot. 
Thorsten

**Comment by schoebel** [comment]
Hi Thorsten,

for an intree build of the current stable branch light0.1.y you will
need _two_ patches: one of the both pre-patches (depending on kernel
version), and the "activation" patch. For an external build, only the
pre-patch is needed.

There is an experimental branch WIP-compatibility which is based on the
0.2 series, and aims to run without the pre-patch. It has lower
performance because it currently cannot use AIO. The current version
appears ok with kernels up to 4.2, but there may be bugs with kernels
4.3 and 4.4.

I will push a new version with some fixes to this branch in the next
days. It will automatically detect the presence of the pre-patch
(whether MARS_MAJOR is defined or not) and use AIO when possible,
otherwise use SIO instead.

Is there a demand for having the same portability with the 0.1 stable
series? Or is 0.2 emough, once it becomes stable?

In summary, there could be all of the following build methods:

 - intree with prepatch (preferred)

 - intree without prepatch (new, less performance)

 - external with prepatch (preferred)

 - external without prepatch (new, possibly for distros)

Testing of all combinations of build methods with all MARS versions and
with all kernel versions is soon becoming a nightmare. Each run of the
old test-suite takes more than 1 day.

Cheers,

Thomas
